api/pkg/apps/cms/case_test.go

```diff
@@ -0,0 +1,165 @@
+// +build integration
+
+package cms_test
+
+import (
+	. "github.com/nrc-no/core/pkg/apps/cms"
+	"github.com/stretchr/testify/assert"
+	"reflect"
+)
+
+func (s *Suite) TestCase() {
+	s.Run("API", func() { s.testCaseAPI() })
+	s.SetupTest()
+	s.Run("List filtering", func() { s.testCaseListFilter() })
+}
+
+func (s *Suite) testCaseAPI() {
+	// Create
+	kase := s.mockCases(1)[0]
+	created, err := s.client.Cases().Create(s.ctx, kase)
+	if !assert.NoError(s.T(), err) {
+		s.T().FailNow()
+	}
+	assert.Equal(s.T(), kase, created)
+
+	// GET
+	get, err := s.client.Cases().Get(s.ctx, created.ID)
+	if !assert.NoError(s.T(), err) {
+		s.T().FailNow()
+	}
+	assert.Equal(s.T(), created, get)
+
+	// UPDATE
+	kase.Done = true
+	kase.FormData = &CaseTemplate{FormElements: []CaseTemplateFormElement{{
+		Type: "textarea",
+		Attributes: CaseTemplateFormElementAttribute{
+			Label: "mock",
+			Value: []string{"mock"},
+		},
+	}}}
+	updated, err := s.client.Cases().Update(s.ctx, kase)
+	if !assert.NoError(s.T(), err) {
+		s.T().FailNow()
+	}
+	assert.Equal(s.T(), kase, updated)
+
+	// GET
+	get, err = s.client.Cases().Get(s.ctx, updated.ID)
+	if !assert.NoError(s.T(), err) {
+		s.T().FailNow()
+	}
+	assert.Equal(s.T(), updated, get)
+
+	// LIST
+	list, err := s.client.Cases().List(s.ctx, CaseListOptions{})
+	if !assert.NoError(s.T(), err) {
+		s.T().FailNow()
+	}
+	assert.Contains(s.T(), list.Items, get)
+}
+
+func (s *Suite) testCaseListFilter() {
+	nCases := 20
+	const (
+		nParties   = 5
+		nTeams     = 4
+		nCaseTypes = 5
+		nParents   = 30
+	)
+
+	parties := []string{}
+	teams := []string{}
+	caseTypes := []string{}
+	parents := []string{}
+
+	cases := s.mockCases(nCases)
+	for i := 0; i < nParties; i++ {
+		parties = append(parties, newUUID())
+	}
+	for i := 0; i < nTeams; i++ {
+		teams = append(teams, newUUID())
+	}
+	for i := 0; i < nCaseTypes; i++ {
+		caseTypes = append(caseTypes, newUUID())
+	}
+	for i := 0; i < nParents; i++ {
+		parents = append(parents, newUUID())
+	}
+
+	// Prepare test data
+	for i, c := range cases {
+		c.PartyID = parties[i%len(parties)]
+		c.TeamID = teams[i%len(teams)]
+		c.CaseTypeID = caseTypes[i%len(caseTypes)]
+		c.ParentID = parents[i%len(parents)]
+		c.Done = i%2 == 0
+		_, err := s.client.Cases().Create(s.ctx, c)
+		if err != nil {
+			s.T().FailNow()
+		}
+	}
+
+	s.Run("by party type", func() { s.testCaseFilterBy("PartyIDs")(cases, parties) })
+	s.Run("by team", func() { s.testCaseFilterBy("TeamIDs")(cases, teams) })
+	s.Run("by case type", func() { s.testCaseFilterBy("CaseTypeIDs")(cases, caseTypes) })
+	s.Run("by parent", func() { s.testCaseFilterBy("ParentID")(cases, parents) })
+	s.Run("by done", func() { s.testCaseFilterByDone(cases) })
```

Comment:
When you run subtests, you need to pass the `testing.T` to the submethod, otherwise you'll run into some weird stuff

```
s.T().Run({NAME}, func(t *testing.T){
   ... at this point, do not use `s.T()`, but rather `t`
})
s.T().Run("by party type", func(t *testing.T) { s.testCaseFilterBy("PartyIDs")(t, cases, parties) })

```

Reply:
if I understand this code snippet from the testify docs, it's actually the other way around:
```golang
	for _, t := range []struct {
		testName string
	}{
		{"first"},
		{"second"},
	} {
		suiteT := suite.T()
		suite.Run(t.testName, func() {
			// We should get a different *testing.T for subtests, so that
			// go test recognizes them as proper subtests for output formatting
			// and running individual subtests
			subTestT := suite.T()
			suite.NotEqual(subTestT, suiteT)
		})
		suite.Equal(suiteT, suite.T())
	}
```

Reply:
Tests run fine btw